shadowsocks/local.go

```diff
@@ -112,34 +83,37 @@ func ListenLocalTCPOptions(options *ListenerOptions) net.Listener {
 		validator = onet.RequirePublicIP
 	}
 
-	isLocal := options.ShouldHandleLocally
-	if isLocal == nil {
-		isLocal = AlwaysLocal
-	}
+	authFunc := service.NewShadowsocksStreamAuthenticator(options.Ciphers, options.ReplayCache, options.ShadowsocksMetrics)
+	tcpHandler := service.NewTCPHandler(options.Listener.Addr().(*net.TCPAddr).Port, authFunc, options.ShadowsocksMetrics, timeout)
+	tcpHandler.SetTargetDialer(&LocalDialer{connections: l.connections})
+
+	accept := func() (transport.StreamConn, error) {
+		listener, ok := l.wrapped.(*tcpListenerAdapter)
+		if !ok {
+			return nil, errors.New("wrapped listener is not a tcpListenerAdapter")
+		}
 
-	dialer := maybeLocalDialer(isLocal, l.dialPipe, service.DefaultDialTarget)
-	l.TCPService = service.NewTCPService(
-		options.Ciphers,
-		options.ReplayCache,
-		&metrics.NoOpMetrics{},
-		timeout,
-		&service.TCPServiceOptions{
-			DialTarget:        dialer,
-			TargetIPValidator: validator,
-		},
-	)
-
-	go func() {
-		err := l.Serve(options.Listener)
-		if err != nil {
-			log.Errorf("serving on %s: %v", l.Addr(), err)
+		conn, err := listener.AcceptTCP()
+		if err == nil {
+			conn.SetKeepAlive(true)
 		}
-		l.Close()
-	}()
 
+		return conn, err
+	}
+
+	handler := func(ctx context.Context, conn transport.StreamConn) {
+		// Add the client connection to the context so it can be used by the LocalDialer
+		ctx = context.WithValue(ctx, ClientConnCtxKey{}, conn)
+		tcpHandler.Handle(ctx, conn)
+	}
+
+	go service.StreamServe(accept, handler)
 	return l
 }
 
+// ClientConnCtxKey is a context key being used to share the client connection
+type ClientConnCtxKey struct{}
```

Comment:
Nit: this doesn't need to be exported, right?  We should refrain from exporting things unless necessary.